### ISSUE TITLE
Run large scale diffs of eyecite output on every PR

### ISSUE THREAD

**Issue by jcushman** [freelawproject/eyecite#117]
A proposal: I think it would help a lot if we had a Github Action that showed the effects of a pull request on a large corpus of random cases. I would script it something like this:

- `benchmark.py corpus.zip ref1 ref2` compares the eyecite codebase at ref1 and ref2 when run against the zipped folder of text files in corpus.zip.
  - By default ref1 is `main` and ref2 is `HEAD`, so it compares upstream with the current code.
- We could have a few standard corpus_10MB.zip, corpus_100MB.zip type files stored as github releases, sampled evenly from documents we care about. I might have a size that takes about 30 seconds to process for local iteration, and a size that takes about 5 minutes to process for CI?
- for each of ref1 and ref2, the script clones the repo into a temp dir; makes a temp python venv and installs dependencies; extracts citations from benchmark_cases.zip; and writes the resulting citations to a json file
- finally the script stores a diff of the two files and reports the size of the diff and how long each extraction took

This could be very useful locally, and also could be wired up into a Github Action that ran `benchmark.py corpus_100MB.zip` and saved the results as [artifacts](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts) of the run, and reported the runtime and diff size as a PR comment.

I'd anticipate two benefits of this:

- For bigger PRs like [partial cite extraction](https://github.com/freelawproject/eyecite/pull/116) where we don't know how courts actually cite things, it would be easier to see whether our guess was right.
- For smaller PRs where we think we aren't changing much, it would stand out if we were actually adding a bunch of false positives or negatives or slowing down performance.

Basically for a project that depends heavily on heuristics, I think it would be great to have more of a sense of whether a heuristic tweak is helping or hurting on average.

**Comment by mlissner** [comment]
Yeah, this is a really good idea. I'd be pretty excited if somebody wanted to run with this. It'd remove some of the uncertainty we have around the little tweaks we make here and there. 

I've seen some other tools that seem to keep track of regressions in things like coverage, I think. Might be nice to be able to say, "Speed is down by 10% and 20% more citations were found." I assume that requires some sort of DB that isn't built into Github, but I've never looked to see if there are simpler solutions for this. 

Anyhow, yeah, this feels like a good idea to me.

**Comment by jcushman** [comment]
> I assume that requires some sort of DB that isn't built into Github, but I've never looked to see if there are simpler solutions for this.

What might work is to download an artifact from the last successful run on the main branch using [this action](https://github.com/dawidd6/action-download-artifact), and compare it to the same artifact generated in the current run.

Seems like one could start just with a local benchmark script to dump citations and timing information to json from a zip file, and iterate on CI integration from there.

**Comment by flooie** [comment]
This has been implemented @jcushman.  

Although it could certainly and will certainly improve over time, any PR in reporters-db or eyecite will trigger a comparison of the PRs against the current eyecite/reporters and give a sample of improvements or degradations.  

We have some trickiness with the token, currently its using one of mine, so I am going to leave this open until we update the token to the `freelawbot_token`

**Comment by jcushman** [comment]
This is awesome -- thanks, Bill!
